File: src/main/Java/boj_step19/boj10872.java
import java.io.*;

public class boj10872 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        writer.append(factorial(n)).close();;
    }
    private static String factorial(int n) {
        return factorialTail(n, 1);
    }
    private static String factorialTail(int n, int acc) {
        if (n == 1) return String.valueOf(acc);
        return factorialTail(n - 1, acc * n);
    }
}

File: src/main/boj_step19/boj10872.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    w.write(factorial(n).toString())
    w.close()
}

fun factorial(n: Int): Int {
    if (n == 0) return 1
    return factorialTail(n, 1)
}

fun factorialTail(n: Int, acc: Int): Int {
    if (n == 1) return acc
    return factorialTail(n - 1, acc * n)
}